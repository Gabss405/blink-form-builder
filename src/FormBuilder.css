.app-container {
  display: flex;
  flex-direction: column;
  min-height: 60vh;
  font-size: 2rem;
  color: white;
}

.form-header {
  background-color: #817091;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: fit-content;
  min-width: 20vw;
  font-size: 2.5rem;
  padding: 10px 50px 10px 50px;
  color: white;
  border-radius: 30px 0px 30px 30px;
}

.form-body {
  min-height: 60vh;
}
.form-description {
  font-style: italic;
  font-size: 2rem;
}

.form-query-select,
.form-date-select,
.form-query-body,
.btn-submit,
.btn-disabled {
  border: 2px solid #817091;
  border-radius: 6px;
  padding: 10px;
  background-color: transparent;
  color: white;
  font-size: 1.5rem;
}

.form-query-select-default {
  opacity: 0.5;
}

.form-query-body {
  width: 100%;
  min-height: 20vh;
}

.form-query-body::placeholder {
  color: white;
  opacity: 0.5;
  font-style: italic;
}

.btn-container {
  display: flex;
  justify-content: center;
  padding: 10px;
}
.btn-submit {
  background-color: white;
  color: #817091;
}

.btn-disabled {
  opacity: 0.5;
  font-size: 1.5rem;
}

.form-query-select:focus-visible,
.form-date-select:focus-visible,
.form-query-body:focus-visible {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none;
  -ms-appearance: none;
}

.btn-submit:hover {
  background-color: #80b3e6;
  cursor: pointer;
}
.btn-submit:active {
  background-color: white;
}
