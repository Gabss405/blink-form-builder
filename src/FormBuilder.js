import { useEffect, useState } from "react";
import "./FormBuilder.css";
const payrollQueryForm = require("./assets/payrollEnquiry.json");

function FormBuilder() {
  const [newPayrollQuery, setNewPayrollQuery] = useState({
    queryType: "",
    queryDate: "",
    queryBody: "",
  });
  const [isFormValid, setIsFormValid] = useState(false);

  const mockUserEmail = "mail@example.org";

  const handleFormChange = (e) => {
    const { name, value } = e.target;
    setNewPayrollQuery({ ...newPayrollQuery, [name]: value });
  };

  const handleFormSubmit = (e) => {
    e.preventDefault();
    //wasn't sure what this means: "the submit button should log an html string of an email body that could be sent to the form's recipient"
    window.open(
      `mailto:${mockUserEmail}?subject=${
        payrollQueryForm.title + " - " + newPayrollQuery.queryType
      }&body=${newPayrollQuery.queryBody}`,
      "_blank"
    );
    console.log(`<body>${newPayrollQuery.queryBody}</body>`);
  };

  useEffect(() => {
    const queryTypeRequiresDate =
      newPayrollQuery.queryType === "Missing Expense" ||
      newPayrollQuery.queryType === "Incorrect Pay";

    const queryDateAndBodyValid =
      (newPayrollQuery.queryDate.length > 0) &
      (newPayrollQuery.queryBody.length > 0);

    const queryTypeNoDateRequired =
      !queryTypeRequiresDate & (newPayrollQuery.queryBody.length > 0);

    if (queryTypeRequiresDate & queryDateAndBodyValid) setIsFormValid(true);
    else if (queryTypeNoDateRequired) setIsFormValid(true);
    else setIsFormValid(false);
  }, [newPayrollQuery]);

  return (
    <div className="app-container">
      <header className="form-header" data-testid="header">
        {payrollQueryForm.title ? payrollQueryForm.title : "Untitled Form"}
      </header>
      <form className="form-body" onSubmit={handleFormSubmit}>
        <p className="form-description" data-testid="form-description">
          {payrollQueryForm?.description && payrollQueryForm.description}
        </p>
        <p className="form-query-type-label">
          {payrollQueryForm?.properties.queryType.description}
        </p>
        <select
          className="form-query-select"
          name="queryType"
          onChange={handleFormChange}
          data-testid="queryType"
          required={payrollQueryForm.required.includes("queryType")}
        >
          <option className="form-query-select-default" value="">
            Select an option
          </option>
          {payrollQueryForm.properties.queryType.items.map((item, idx) => {
            return (
              <option key={idx} value={item}>
                {item}
              </option>
            );
          })}
        </select>
        <p>{payrollQueryForm.properties.date.title}</p>
        <input
          className="form-date-select"
          type="date"
          name="queryDate"
          onChange={handleFormChange}
          data-testid="queryDate"
          required={
            newPayrollQuery.queryType === "Missing Expense" ||
            newPayrollQuery.queryType === "Incorrect Pay"
          }
        ></input>
        <p>Query</p>
        <textarea
          className="form-query-body"
          type="text"
          name="queryBody"
          data-testid="queryBody"
          value={newPayrollQuery.queryBody}
          onChange={handleFormChange}
          placeholder="Enter more information here..."
          required={payrollQueryForm.required.includes("queryBody")}
        ></textarea>
        <div className="btn-container">
          {isFormValid ? (
            <button type="submit" data-testid="submit" className="btn-submit">
              Submit
            </button>
          ) : (
            <button
              type="submit"
              data-testid="submit"
              disabled
              className="btn-disabled"
            >
              Submit
            </button>
          )}
        </div>
      </form>
    </div>
  );
}

export default FormBuilder;
