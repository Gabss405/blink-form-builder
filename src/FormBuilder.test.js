import FormBuilder from "./FormBuilder";
import { render, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
const payrollQueryForm = require("./assets/payrollEnquiry.json");

beforeEach(cleanup);

describe("<FormBuilder> rendering", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(<FormBuilder />);
    expect(asFragment(<FormBuilder />)).toMatchSnapshot();
  });
  it("renders header, main description, query type and date title from JSON", () => {
    const { getByTestId, getByText } = render(<FormBuilder />);
    expect(getByTestId("header")).toHaveTextContent(payrollQueryForm.title);
    expect(getByText(payrollQueryForm.description)).toBeInTheDocument();
    expect(
      getByText(payrollQueryForm.properties.queryType.description)
    ).toBeInTheDocument();
    expect(
      getByText(payrollQueryForm.properties.date.title)
    ).toBeInTheDocument();
  });
  it("requires same fields as in JSON document ", () => {
    const { getByTestId } = render(<FormBuilder />);
    payrollQueryForm.required.forEach((item) => {
      expect(getByTestId(item)).toBeRequired();
    });
  });
  it("submit button should be disabled by default", () => {
    const { getByTestId } = render(<FormBuilder />);
    expect(getByTestId("submit")).toBeDisabled();
  });
});

describe("<FormBuilder> validation", () => {
  it("renders submit button disabled if fields are required", async () => {
    const { getByTestId, getByText } = render(<FormBuilder />);

    userEvent.selectOptions(getByTestId("queryType"), [
      getByText("Missing Expense"),
    ]);
    userEvent.type(getByTestId("queryBody"), "Hello World");
    expect(getByTestId("submit")).toBeDisabled();

    userEvent.selectOptions(getByTestId("queryType"), [
      getByText("Change of Bank Details"),
    ]);
    expect(getByTestId("submit")).not.toBeDisabled();
  });
});

describe("<FormBuilder> submission", () => {
  it("submits form if form values are valid ", async () => {
    const { getByTestId, getByText } = render(<FormBuilder />);
    console.log = jest.fn();
    window.open = jest.fn();

    userEvent.selectOptions(getByTestId("queryType"), [
      getByText("Change of Address"),
    ]);
    userEvent.click(getByTestId("queryBody"));
    userEvent.type(getByTestId("queryBody"), "Hello World");
    userEvent.click(getByTestId("submit"));

    expect(getByTestId("submit")).not.toBeDisabled();
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledWith(`<body>Hello World</body>`);
    expect(window.open).toHaveBeenCalledTimes(1);
    expect(window.open).toHaveBeenCalledWith(
      `mailto:mail@example.org?subject=${
        payrollQueryForm.title + " - " + "Change of Address"
      }&body=Hello World`,
      "_blank"
    );
  });
});
