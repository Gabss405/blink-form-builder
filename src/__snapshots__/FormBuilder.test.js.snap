// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormBuilder> rendering should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="app-container"
  >
    <header
      class="form-header"
      data-testid="header"
    >
      Payroll Enquiry
    </header>
    <form
      class="form-body"
    >
      <p
        class="form-description"
        data-testid="form-description"
      >
        Please use this form for any Payroll related enquiries.
      </p>
      <p
        class="form-query-type-label"
      >
        What does your query relate to?
      </p>
      <select
        class="form-query-select"
        data-testid="queryType"
        name="queryType"
        required=""
      >
        <option
          class="form-query-select-default"
          value=""
        >
          Select an option
        </option>
        <option
          value="Incorrect Pay"
        >
          Incorrect Pay
        </option>
        <option
          value="Missing Expense"
        >
          Missing Expense
        </option>
        <option
          value="Change of Bank Details"
        >
          Change of Bank Details
        </option>
        <option
          value="Change of Address"
        >
          Change of Address
        </option>
        <option
          value="Other"
        >
          Other
        </option>
      </select>
      <p>
        Date of payslip being queried (if applicable)
      </p>
      <input
        class="form-date-select"
        data-testid="queryDate"
        name="queryDate"
        type="date"
      />
      <p>
        Query
      </p>
      <textarea
        class="form-query-body"
        data-testid="queryBody"
        name="queryBody"
        placeholder="Enter more information here..."
        required=""
        type="text"
      />
      <div
        class="btn-container"
      >
        <button
          class="btn-disabled"
          data-testid="submit"
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
